### PR TITLE
Improve content offer

### PR DESCRIPTION
**Pull request by KonradStaniec**
Modifies `offer` so that it is possible to provide `content: seq[byte]` directly instead of getting it from database. 
Leaves possibility of offering content from database as it have its use cases.
From docs added to `offer` proc:
```
  ## Direct - when caller provides content to transfer. This way, content is 
  ## guaranteed to be transferred as it stays in memory until whole transfer
  ## is completed.
  ## Database - when caller provides keys of content to be transferred. This 
  ## way content is provided from database just before it is transferred through
  ## uTP socket. This is useful when there is a lot of content to be transferred
  ## to many peers, and keeping it all in memory could exhaust node resources.
  ## Main drawback is that content maybe deleted from node database 
  ## by cleanup process before it will be transferred, so this way does not 
  ## guarantee content transfer.  
```

### CODE REVIEW

File: fluffy/network/wire/portal_protocol.nim [hunk 1]
```diff
@@ -588,75 +612,121 @@ proc findContent*(p: PortalProtocol, dst: Node, contentKey: ByteList):
...
-    var requestedContentKeys: seq[ByteList]
-    for i, b in m.contentKeys:
-      if b:
-        requestedContentKeys.add(contentKeys[i])
-
-    let contentKeysAmount = requestedContentKeys.len()
-    portal_content_keys_accepted.observe(contentKeysAmount.int64)
-    if contentKeysAmount == 0:
-      # Don't open an uTP stream if no content was requested
+    ## Offer triggers offer-accept interaction with one peer
```
Comment (kdeme) [inline fluffy/network/wire/portal_protocol.nim:627]: The indent on all this is one off

File: fluffy/network/wire/portal_protocol.nim [hunk 2]
```diff
@@ -588,75 +612,121 @@ proc findContent*(p: PortalProtocol, dst: Node, contentKey: ByteList):
...
+proc offer*(p: PortalProtocol, dst: Node, contentKeys: ContentKeysList):
+    Future[PortalResult[void]] {.async.} =
+    let req = OfferRequest(dst: dst, kind: Database, contentKeys: contentKeys)
+    let res = await p.offer(req)
+    return res
+
+proc offer*(p: PortalProtocol, dst: Node, content: seq[ContentInfo]):
+    Future[PortalResult[void]] {.async.} =
+    if len(content) > contentKeysLimit:
+      return err("Cannot offer more that 64 content items")
```
Comment (kdeme) [inline fluffy/network/wire/portal_protocol.nim:718]:
```suggestion
      return err("Cannot offer more than 64 content items")
```

**Review by kdeme — APPROVED** [review]

File: fluffy/network/history/history_network.nim [hunk 3]
```diff
@@ -132,20 +132,18 @@ proc getBlockHeader*(
...
-  if maybeHeader.isSome() and h.portalProtocol.inRange(contentId):
-    # TODO this bit is quite troubling, currently we may trigger offer/accept
-    # only when content is in our db and we save content only when is in our range
-    # which means we cannot propagate content which is not in our range, but maybe
-    # in range of other nodes. 
-    h.contentDB.put(contentId, headerContent.content)
-    # content is valid and in the db, it may be propagated it through the network
+  if maybeHeader.isSome():
+    # Content is valid we can propagate it to interested peers
     h.portalProtocol.triggerPoke(
```
[the remark this answers is not in the document]
Comment (KonradStaniec) [reply fluffy/network/history/history_network.nim:137]: currently what stopping us is content validation, as it is only done at history network level. The way around is probably making portal wire protocol accepting some kinda callback to handle and validate content. It requires either way to validate content we propagate as part of neighbour gossip. Imo it should be addressed in separate pr.

File: fluffy/network/wire/portal_protocol.nim [hunk 4]
```diff
@@ -120,6 +120,21 @@ type
...
+    content*: seq[byte]
+
+  OfferRequestType = enum
+    Direct, Database
+
+  OfferRequest = object
+    dst: Node
+    case kind: OfferRequestType
+    of Direct:
+      contentList: List[ContentInfo, contentKeysLimit]
```
[the remark this answers is not in the document]
Comment (KonradStaniec) [reply fluffy/network/wire/portal_protocol.nim:135]: oh it is definitly good to know, as currently I kinda treated `List[Type, maxLenght]` as collection which makes sure we do not go over the limit.

Although I would still leave it here and push limit checking to the edges as having it here tells reader of the code quite much about how long collection we are expecting here.

File: fluffy/network/history/history_network.nim (hunk 3 above)
[the remark this answers is not in the document]
Comment (kdeme) [reply fluffy/network/history/history_network.nim:137]: Right, I'm actually adding that callback to do the content validation on received data, as I've already had some situations of invalid data in the network, and it is hard to get out again when that occurs :).

**Pull request merged**